plugins/database/mongodb/connection_producer.go

```diff
@@ -138,10 +138,7 @@ func (c *mongoDBConnectionProducer) Connection(_ context.Context) (interface{},
 	return c.session, nil
 }
 
-// Close terminates the database connection.
 func (c *mongoDBConnectionProducer) Close() error {
-	c.Lock()
```

Comment:
Locking inside Close() always produces a deadlock, as all of the calling functions (from mongodb.go) have locked the mutex already.

Reply:
Close can be called externally which is why we need to lock.

Reply:
Alright, I see two possibilities for dealing with that then.

1) Unlocking before calling it, and then locking again once done. (downside being that the lock needs to be acquired twice, and that another process can interfere inbetween the locking and unlocking)

2) Not calling it at all, and just closing the session by referencing the mongoDBConnectionProducer in mongodb.go. I had a commit where I did exactly this, and it worked ok.

What do you prefer?

Reply:
@calvn Close now locks again!